ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/BlockingPrintPublisher.java

```diff
@@ -0,0 +1,203 @@
+  private static final Duration POLL_TIMEOUT = Duration.ofMillis(5000L);
+  public static final int SEND_MAX_BATCH_SIZE = 200;
+
+  private final WorkerExecutor workerExecutor;
+  private final KafkaConsumer<Bytes, Bytes> topicConsumer;
+  private final PrintTopic printTopic;
+  private final int limit;
+  private final long interval;
+  private final RecordFormatter formatter;
+  private volatile boolean closed;
+  private volatile boolean started;
+
+  public BlockingPrintPublisher(final Context ctx,
+      final WorkerExecutor workerExecutor,
+      final ServiceContext serviceContext,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> consumerProperties,
+      final PrintTopic printTopic) {
+    super(ctx);
+    this.workerExecutor = Objects.requireNonNull(workerExecutor);
+    this.printTopic = printTopic;
+    this.limit = printTopic.getLimit().orElse(Integer.MAX_VALUE);
+    this.interval = printTopic.getIntervalValue();
+    this.formatter = new RecordFormatter(
+        serviceContext.getSchemaRegistryClient(),
+        printTopic.getTopic()
+    );
+    this.topicConsumer = createTopicConsumer(serviceContext, ksqlConfig, consumerProperties,
+        printTopic);
+  }
+
+  public Future<Void> close() {
+    if (closed) {
+      return Future.succeededFuture();
+    }
+    closed = true;
+    // Run async as it can block
+    ctx.runOnContext(v -> sendComplete());
+
+    return super.close();
+  }
+
+
+  @Override
+  protected void maybeSend() {
+    executeOnWorker(this::doSend);
+  }
+
+  @Override
+  protected void afterSubscribe() {
+    if (!started) {
+      started = true;
+    }
+  }
+
+  public void startFromWorkerThread() {
+    VertxUtils.checkIsWorker();
+    started = true;
+  }
+
+  private void executeOnWorker(final Runnable runnable) {
+    workerExecutor.executeBlocking(p -> runnable.run(), false, ar -> {
+      if (ar.failed()) {
+        log.error("Failed to close print", ar.cause());
+      }
+    });
+  }
+
+  private void doSend() {
+    VertxUtils.checkIsWorker();
+    int messagesWritten = 0;
+    int messagesPolled = 0;
+
+    while (!isClosed() && getDemand() > 0 && messagesWritten < limit) {
+      final ConsumerRecords<Bytes, Bytes> records = topicConsumer.poll(POLL_TIMEOUT);
+      if (records.isEmpty()) {
+        continue;
+      }
+      final List<String> formattedRecords = formatter.format(
+          records.records(this.printTopic.getTopic())
+      );
+
+      for (final String message : formattedRecords) {
+        if (messagesPolled++ % interval == 0) {
+          messagesWritten++;
+          doOnNext(message);
+        }
+        if (messagesWritten >= limit) {
+          ctx.runOnContext(v -> sendComplete());
+          break;
+        } else if (messagesWritten >= SEND_MAX_BATCH_SIZE) {
+          ctx.runOnContext(v -> doSend());
+          break;
+        }
+      }
+    }
+
+    if (isClosed()) {
+      topicConsumer.close();
```

Comment:
Why did this get moved out of `close()` to here instead? Does this still guarantee that the topic consumer is closed when the print publisher is?

Reply:
Yes, the topic consumer is closed when the publisher is. I've tested this a couple of times.

I moved this because of an error/warning from kafkaConsumer:` KafkaConsumer is not safe for multi-threaded access`. Basically, poll and close functions were being called from different threads and kafkaConsumer and this is not safe. That's fixed now :)